### ISSUE TITLE
Not working if function returns None

### ISSUE THREAD

**Issue by AckslD** [python-cachier/cachier#63]
I just tried out `cachier` for the first run and couldn't at first understand why it was not working and the reason seems to be that caching is not done if the function returns None. Is this intended?

Consider the following example:
```python
from time import sleep
from timeit import default_timer as timer

from cachier import cachier


@cachier()
def func():
    sleep(1)


t0 = timer()
func()
t1 = timer()
func()
t2 = timer()
print(t1 - t0)
print(t2 - t1)
```
where no caching is done.
However, changing the function to be:
```python
@cachier()
def func():
    sleep(1)
    return 1
```
enables the caching

**Comment by shaypal5** [comment]
Well, yes, it is intended, in the sense that there's no use in decorating a function with the `@cachier` decorator if it does not return anything - the whole point of the decorator is to cache return values.

Please refer the documentation for additional information. :)

**Comment by AckslD** [comment]
@shaypal5 I understand it's no point in caching a function that only returns `None` but what if the function returns for example `Optional[int]`?

**Comment by shaypal5** [comment]
A function can't return `Optional[int]` - it is not a type, it is a type hint. Specifically, it's a shorthand for `Union[int, None]`.
Concretely, in any specific call the function will return either a `None` or an `int`.

I'm guessing that when it returns an `int` the results will be cached. I'm not sure what happens if you explicitly `return None`.
I suggest you build a specific use case where you will want both `int` and an explicit `None`, check what happens and get back to me.

If I find out there's a real use case not being served, we can open a new issue that suggests an enhancement (it *will* wait for someone to implement it, however; this is a small personal project, so it takes someone else taking initiative for most enhancement).

**Comment by AckslD** [comment]
I'm just trying to raise a point, but that's alright. I'll open another issue if it's stopping me to do something :)

**Comment by jall** [comment]
@shaypal5 I can confirm `cachier` v2.1.1 doesn't cache if you explicitly `return None`

This seems like strange behaviour to me: `None` is a value, and it seems like a common pattern to do something like this

```python
def get_user(id: str) -> Optional[User]:
   response = request.get(f"/users/{id}")
   if response.status_code == 404:
      return None
   return response.body
```

In my case, I have a script with thousands of records that need to do an http lookup for a possible match (in a smallish set of possibilities). I'd like to cache these to save doing an http request hundreds of times instead of once, and None is a valid return value here

There's a workaround where I can return `User | Literal["None"]` but this is a bit messy

**Comment by shaypal5** [comment]
I get it. I agree that it sounds like a useful behavior, and a sensible one at that. So yeah, let's promote this to a feature request.

**Comment by jall** [comment]
> One possible issue I can see is that any code outside a function has no way to tell whether `None` was returned explicitly or not.

@shaypal5 I don't know of a method to inspect this from outside, short of ast parsing for a `return None`, so in #121 I opted to add another config param which can be set per-function or globally

The default behaviour remains the same, so it's up to the caller to decide if the function should be cached or not

**Comment by AckslD** [comment]
My 2 cents :)

I don't see why not just treat `None` as any other hashable value? Same as `functools.cache`. So instead of doing eg (assuming `entry` is a `dict`):
```python
if entry.get('value') is None: ...
```
(side-point, note that you don't need to pass `None` to `get`, that's the default default.)
do
```python
if 'value' in entry: ...
```
Which would allow `entry` having a `'value'` which is `None`.

That said, even if you want to treat `None` specially I definitely think one shouldn't distinguish:
```python
def f():
    return None
```
from
```python
def f():
    pass
```
They are the same thing and both return `None` as the value. Having any code which distinguishes between explicit `return None` vs implicit is I think asking for trouble.

**Comment by lordjabez** [comment]
I agree with @AckslD here, it seems `cachier` should be able to support a return value of `None` without another configuration option. Unfortunately it's not quite as simple as changing the line mentioned, as there are portions of the code that assume `value` being in `entry` but equal to `None` means a value doesn't yet exist.

I'm confident that can be addressed but will take some careful work.

**Comment by shaypal5** [comment]
Just updating we're close to pulling in a PR by Jon (@jall) that should solve this. Just getting through some much needed refactors and freshing ups of the tests.

**Comment by shaypal5** [comment]
Done by @jall and released in `v2.2.0`!🎉
